src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,87 +77,295 @@ def __eq__(self, other: Any) -> bool:
+        """Falcon mask.
+
+        Args:
+            x_sh (ReplicatedSharedTensor): X share
+            y_sh (ReplicatedSharedTensor) : Y share
+            op_str (str): Operator
+
+        Returns:
+            values(Tuple[ReplicatedSharedTensor,ReplicatedSharedTensor]) : masked_values.
+        """
+        session = get_session(x_sh.session_uuid)
+
+        crypto_store = session.crypto_store
+
+        primitives = crypto_store.get_primitives_from_store(
+            f"beaver_{op_str}", x_sh.shape, y_sh.shape, remove=False
+        )
+
+        a_sh, b_sh, _ = primitives
+
+        return x_sh - a_sh, y_sh - b_sh
+
+    @staticmethod
+    def mul_malicious(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        **kwargs_: Dict[Any, Any],
+    ) -> MPCTensor:
+        """Falcon malicious multiplication.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            kwargs_ (Dict[Any, Any]): Kwargs for some operations like conv2d
+
+        Returns:
+            result(MPCTensor): Result of the operation.
+
+        Raises:
+            ValueError : If the shares are not valid.
+        """
+        shape_x = tuple(x.shape)
+        shape_y = tuple(y.shape)
+
+        result = Falcon.mul_semi_honest(
+            x, y, session, op_str, truncate=False, **kwargs_
+        )
+
+        args = [list(sh) + [op_str] for sh in zip(x.share_ptrs, y.share_ptrs)]
+        try:
+            mask = parallel_execution(Falcon.falcon_mask, session.parties)(args)
+        except EmptyPrimitiveStore:
+            CryptoPrimitiveProvider.generate_primitives(
+                f"beaver_{op_str}",
+                session=session,
+                g_kwargs={
+                    "session": session,
+                    "a_shape": shape_x,
+                    "b_shape": shape_y,
+                    "nr_parties": session.nr_parties,
+                    **kwargs_,
+                },
+                p_kwargs={"a_shape": shape_x, "b_shape": shape_y},
+            )
+            mask = parallel_execution(Falcon.falcon_mask, session.parties)(args)
+
+        eps_shares, delta_shares = zip(*mask)
+
+        eps = MPCTensor(shares=eps_shares, session=session)
+        delta = MPCTensor(shares=delta_shares, session=session)
+
+        eps_plaintext = eps.reconstruct(decode=False)
+        delta_plaintext = delta.reconstruct(decode=False)
+
+        args = [
+            list(sh) + [eps_plaintext, delta_plaintext, op_str]
+            for sh in zip(result.share_ptrs)
+        ]
+
+        triple_shares = parallel_execution(Falcon.triple_verification, session.parties)(
+            args, kwargs_
+        )
+
+        triple = MPCTensor(shares=triple_shares, session=x.session)
+
+        if (triple.reconstruct(decode=False) == result.reconstruct(decode=False)).all():
```

Comment:
Q: Wouldn't be better at this step if we don't reconstruct the result of the multiplication. but reconstruct this
![image](https://user-images.githubusercontent.com/7805588/124677325-1d308580-deb8-11eb-8d37-6db40973ccc2.png)